pkg/controller/ipam/antrea_ipam_controller.go

```diff
@@ -0,0 +1,243 @@
+		namespaceListerSynced:   namespaceInformer.Informer().HasSynced,
+		podLister:               podInformer.Lister(),
+		podListerSynced:         podInformer.Informer().HasSynced,
+		statefulSetInformer:     statefulSetInformer,
+		statefulSetListerSynced: statefulSetInformer.Informer().HasSynced,
+		ipPoolLister:            ipPoolInformer.Lister(),
+		ipPoolListerSynced:      ipPoolInformer.Informer().HasSynced,
+	}
+
+	// Add handlers for Stateful Set events.
+	klog.V(2).InfoS("Subscribing for StatefulSet notifications", "controller", controllerName)
+	statefulSetInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			DeleteFunc: c.enqueueStatefulSetDeleteEvent,
+		},
+	)
+	return c
+}
+
+// Enqueue the STatefulSet delete notification to process by the worker
+func (c *AntreaIPAMController) enqueueStatefulSetDeleteEvent(obj interface{}) {
+	ss := obj.(*appsv1.StatefulSet)
+	klog.V(2).InfoS("Delete notification", "StatefulSet", ss.Name)
+
+	key := ss.GetUID()
+	c.statefulSetMap.Store(key, ss)
+	c.statefulSetCleanupQueue.Add(key)
+}
+
+// Find IP Pools annotated to StatefulSet via direct annotation or namespace annotation
+func (c *AntreaIPAMController) getIPPoolsByStatefulSet(ss *appsv1.StatefulSet) []string {
+
+	// Inspect pool annotation for the Pods
+	// In order to avoid extra API call in IPAM driver, IPAM annotations are defined
+	// on Pods rather than on StatefulSet
+	selector := map[string]string{"app": ss.Name}
+	podList, err := c.podLister.Pods(ss.Namespace).List(labels.SelectorFromSet(labels.Set(selector)))
+
+	if err != nil {
+		klog.Errorf("Failed to list Pods for StatefulSet %s/%s", ss.Namespace, ss.Name)
+		return nil
+	}
+
+	for _, pod := range podList {
+		annotations, exists := pod.Annotations[AntreaIPAMAnnotationKey]
+		if exists {
+			// Stateful Set Pod is annotated with dedicated IP pool
+			return strings.Split(annotations, AntreaIPAMAnnotationDelimiter)
+		}
+	}
+
+	// Inspect Namespace
+	namespace, err := c.namespaceLister.Get(ss.Namespace)
+	if err != nil {
+		// Should never happen
+		klog.Errorf("Namespace %s not found for StatefulSet %s", ss.Namespace, ss.Name)
+		return nil
+	}
+
+	annotations, exists := namespace.Annotations[AntreaIPAMAnnotationKey]
+	if exists {
+		return strings.Split(annotations, AntreaIPAMAnnotationDelimiter)
+	}
+
+	return nil
+
+}
+
+// Look for an IP Pool associated with this StatefulSet, either a dedicated one or
+// annotated to the namespace. If IPPool is found, this routine will clear all IP
+// addresses that might be reserved for the pool.
+func (c *AntreaIPAMController) cleanIPPoolForStatefulSet(ss *appsv1.StatefulSet) error {
+	klog.InfoS("Processing delete notification", "StatefulSet", ss.Name)
+	poolNames := c.getIPPoolsByStatefulSet(ss)
+	if len(poolNames) == 0 {
+		// no Antrea IPAM annotation for this StatefulSet
+		return nil
+	}
+
+	klog.InfoS("Clearing reserved IP addresses", "StatefulSet", ss.Name, "IP Pools", poolNames)
```

Comment:
Having two V0 logs for each StatefulSet event seems too verbose.

Reply:
Agreed, removed